### PR TITLE
fix(api): atomic durable-object rate limiting with graceful fallback

### LINKED ISSUE #464: Rate limiter (consumeSharedRateLimit) is non-atomic under concurrency
## Context
Raised by CodeAnt during review of #463.

`app/server/utils/sharedEdgeStore.ts` → `consumeSharedRateLimit` enforces limits via a non-atomic read-then-write against the Cloudflare Cache API (`caches.default`). Under concurrent request bursts — especially across multiple edge isolates, which each keep an independent in-memory fallback map — requests can read a stale `count` and all pass, allowing the effective cap to be exceeded (roughly `limit × isolate count`).

## Affected call sites
- `app/server/api/logs/client.post.ts` (added in #463)
- `app/server/api/tarkov-dev/profile.get.ts`
- `app/server/api/team/members.ts`
- `app/server/api/profile/[userId]/[mode].get.ts`

## Why it wasn't fixed in #463
It's a property of the shared limiter utility, not of that PR's change, and a strict fix is cross-cutting. #463 deliberately stayed scoped to hardening the log sink.

## Proposed fix
The Cloudflare Cache API has no atomic increment/CAS primitive. Options:
- Durable Object token-bucket limiter (strict, recommended for shared utility).
- Cloudflare KV/D1 with compare-and-swap semantics.

## Severity / priority
Low-to-moderate. Current behavior is best-effort and bounds volume rather than enforcing exact caps. Acceptable for fire-and-forget telemetry; worth tightening for the auth/rate-sensitive routes (profile, team).

### PR DESCRIPTION
**Pull request by DysektAI**
## Summary

Closes #464 for the auth/rate-sensitive routes by giving `consumeSharedRateLimit` a globally-atomic enforcement path, while preserving today's behavior everywhere the binding is absent.

The root cause in #464 is that the Cache API has no atomic increment, so the read-then-write counter can be raced across isolates and data centers (worst case ≈ `limit × isolates × colos`). The repo already ships a correct atomic limiter — the SQLite-backed `ApiGatewayRateLimiter` Durable Object in `workers/api-gateway`. This PR reuses it from the Pages deployment instead of standing up new infrastructure.

### Pages-side (atomic limiter wiring)

- **`app/server/utils/sharedEdgeStore.ts`**
  - `consumeSharedRateLimit` now prefers the Durable Object limiter when an `API_GATEWAY_LIMITER` binding is present on the handle. The DO serializes count+increment per key, so enforcement is exact across isolates and locations.
  - On a missing binding **or any DO error/malformed response**, it **fails open** to the existing Cache API + in-memory path — a limiter outage degrades to best-effort rather than breaking requests.
  - DO request is bounded by an `AbortController`-backed 3s timeout so a hung stub falls back to the cache path instead of stalling on Cloudflare's platform timeout.
  - `windowSec` conversion uses `Math.ceil(windowMs / 1000)` so sub-second remainders never produce a shorter-than-configured DO window.
  - Added `getRateLimiterBinding(event)` to read the binding from `event.context.cloudflare.env`, and `resolveSharedRateLimitNamespace` for validation.
  - Documented the enforcement tiers and the precise best-effort limits of the fallback (cross-isolate, cross-colo, in-memory eviction, isolate restarts).
- **Call sites:** wired the binding into the two routes #464 calls out as sensitive — `team/members` and `profile/[userId]/[mode]`. The fire-and-forget client log sink intentionally stays on the best-effort path (a per-request DO hop isn't worth it for telemetry).
- **Pre-auth keying:** team-members and shared-profile pre-auth gates now key on IP only. Embedding attacker-controlled `teamId`/`userId`/`mode` let callers bypass the gate by rotating IDs and spawn unbounded DO instances.
- **`wrangler.toml`:** bind the external DO by `script_name = "api-gateway"` at the top level, mirrored under `[env.preview]` because `durable_objects` is a non-inheritable key (the file already overrides `vars` for preview).

### Worker-side (`ApiGatewayRateLimiter` hardening)

- **Storage cleanup alarm.** Each request schedules a self-cleanup alarm at `resetAt + 1s`. `alarm()` calls `state.storage.deleteAll()` once the window has elapsed, or reschedules to the current window's deadline if a newer window is still active. Alarm comparison uses `existingAlarm !== cleanupAt` so both earlier and later existing alarms are normalized to the current window. Without this, every unique key retained billable SQLite storage forever.

## Why this approach

- Native Rate Limiting binding: not confirmed on Pages Functions, eventually-consistent by design, and config-static (can't express the runtime-configurable team/profile limits). Rejected.
- KV/D1 CAS: KV has no CAS and is limited to one write/key/sec; D1 adds a DB dependency. Rejected.
- Durable Object: the issue's recommended option, already implemented and deployed in this account, accepts `{limit, windowSec}` per call. Chosen.

## Tests

- **`app/server/utils/__tests__/sharedEdgeStore.test.ts`** — DO allowed/denied paths, fail-open on DO error, hung-stub abort + fallback (advances fake timers), and a barrier-based concurrency regression that forces all `cache.match()` calls to observe the same stale value before any write resolves, then asserts a burst of 5 concurrent same-isolate calls yields exactly 3 allowed / 2 denied at `limit=3`. Verified the regression test fails if the in-memory serialization is removed.
- **`workers/api-gateway/src/__tests__/gateway.test.ts`** — alarm scheduling, post-window `deleteAll`, active-window reschedule (asserts `setAlarm` is called with `stored.resetAt + 1000`). 32 tests pass.

Validation: `npm run typecheck`, `npx eslint app --max-warnings=0`, `npm run test:api-gateway`, and the 4 affected endpoint suites all pass.

## Deployment notes / risk

- Cloudflare auto-deploys both Pages and the `api-gateway` Worker on push to `main` / preview branches, so the standard merge flow rolls out the storage-cleanup alarm alongside the Pages binding.
- Because the code fails open, the deploy is safe even if the binding doesn't resolve immediately — endpoints fall back to current behavior.
- `client.post.ts` behavior is unchanged.

Relates to #464.

## Summary by CodeRabbit

* **Infrastructure**
  * Enhanced rate limiting system with improved reliability and enforcement across distributed environments.
  * Updated API endpoints to utilize a more robust rate limiter backed by persistent storage.
  * Improved automatic cleanup of expired rate limit data to prevent stale entries.

* **Tests**
  * Added comprehensive test coverage for rate limiting behavior under concurrent conditions and edge cases.

### CODE REVIEW

**Comment by DysektAI** [comment]
Addressed the P2 in `ea3d769c`. Wrapped the Durable Object `stub.fetch` in a 3s `AbortController` timeout (matching the api-gateway caller's bound). On timeout the request aborts, the catch reports via `onError`, and `consumeDurableRateLimit` returns `null` so the limiter falls open to the Cache API + in-memory path as documented. Added a regression test that advances fake timers to 3s and asserts the hung request aborts and falls back (verified it hangs without the timeout).

File: app/server/utils/sharedEdgeStore.ts [hunk 1]
```diff
@@ -240,6 +274,64 @@ export const writeSharedCache = async <T>(
...
 };
+const consumeDurableRateLimit = async (
+  limiter: SharedRateLimitNamespace,
+  prefix: string,
+  key: string,
+  limit: number,
+  windowMs: number,
+  onError?: SharedCacheErrorHandler
+): Promise<boolean | null> => {
+  const windowSec = Math.max(1, Math.round(windowMs / 1000));
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply app/server/utils/sharedEdgeStore.ts]: Fixed in `ade32f96` — switched to `Math.ceil` so the DO window is never shorter than configured. No behavior change at current call sites (all 60000ms → 60s).

File: app/server/utils/sharedEdgeStore.ts [hunk 2]
```diff
@@ -240,6 +274,64 @@ export const writeSharedCache = async <T>(
...
+  try {
+    const id = limiter.idFromName(`${prefix}:${key}`);
+    const stub = limiter.get(id);
+    const response = await stub.fetch('https://rate-limit', {
+      method: 'POST',
+      headers: { 'content-type': 'application/json' },
+      body: JSON.stringify({ limit, windowSec }),
+    });
+    if (!response.ok) {
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply app/server/utils/sharedEdgeStore.ts:298]: Already addressed in `ea3d769c` (pushed before this review): `consumeDurableRateLimit` now wraps `stub.fetch` in a 3s `AbortController` timeout matching the api-gateway caller. On timeout the request aborts, the catch reports via `onError`, and the function returns `null` so the limiter falls open to the Cache API + in-memory path. Covered by a regression test that advances fake timers to 3s and asserts fallback (verified it hangs without the timeout).

**Comment by DysektAI** [comment]
Addressed the P1 (unbounded DO storage growth) in `5ac05557` with two fixes:

1. **DO self-cleanup** — `ApiGatewayRateLimiter` now schedules a cleanup alarm sized to the rate-limit window. `alarm()` calls `storage.deleteAll()` once the window has elapsed (and reschedules if a newer window is still active), so abandoned keys no longer retain billable SQLite storage indefinitely.
2. **Bounded pre-auth keys** — the unauthenticated rate-limit gate for team-members and shared-profile no longer embeds attacker-controlled `teamId`/`userId`/`mode`. Keying on IP alone stops callers from both bypassing the per-IP limit and spawning unbounded DO instances by rotating IDs.

Added unit tests for alarm scheduling, post-window wipe, and active-window reschedule. No new DO migration needed (adding an `alarm()` handler is a code change; the SQLite class is unchanged). Worker `tsc` + app typecheck/lint clean; api-gateway suite 32 pass, affected app suites 55 pass.
